File: scripts/_shared/nowcast-source-gate.mjs
/**
 * Runtime source-eligibility gate used by inference scripts before a
 * nowcast is written. Returns { abstain: true, reason } when the current
 * set of available sources is too weak to trust.
 *
 * Split into:
 *   - evaluateGate(): pure function. Takes rules, regime, availableSources,
 *                     recentMae, minEligibleSources. Returns abstain decision.
 *   - detectRegime(), checkEligibleSources(): DB-touching wrappers that
 *                     fetch inputs then call evaluateGate. Used by inference.
 */

export const DEFAULT_MIN_ELIGIBLE_SOURCES = 2;

/**
 * Pure gate evaluator. No DB access. Testable with predetermined inputs.
 *
 * @param {{
 *   rules: Array<{ source_signal: string, max_lag_hours: number|string,
 *                  holdout_mae_max: number|string, drift_threshold?: number|string,
 *                  regime_mask?: Record<string, boolean> }>,
 *   regime: 'normal' | 'shock' | 'unknown',
 *   availableSources: Array<{ name: string, lagHours?: number }>,
 *   recentMae?: number | null,
 *   minEligibleSources?: number,
 * }} input
 */
export function evaluateGate({
  rules, regime, availableSources,
  recentMae = null, minEligibleSources = DEFAULT_MIN_ELIGIBLE_SOURCES,
}) {
  if (!Array.isArray(rules) || rules.length === 0) {
    return { abstain: true, eligible: [], rejected: [], regime, reason: 'no eligibility rules' };
  }

  const eligible = [];
  const rejected = [];
  for (const src of availableSources || []) {
    const rule = rules.find((r) => String(r.source_signal) === String(src.name));
    if (!rule) {
      rejected.push({ source: src.name, reason: 'no rule' });
      continue;
    }
    const regimeMask = rule.regime_mask || { normal: true, shock: true };
    if (!regimeMask[regime]) {
      rejected.push({ source: src.name, reason: `regime=${regime} disabled` });
      continue;
    }
    const maxLag = Number(rule.max_lag_hours);
    const srcLag = Number(src.lagHours);
    if (Number.isFinite(srcLag) && Number.isFinite(maxLag) && srcLag > maxLag) {
      rejected.push({ source: src.name, reason: `lag ${srcLag}h > ${maxLag}h` });
      continue;
    }
    eligible.push(src.name);
  }

  let driftExceeded = false;
  let driftReason = null;
  if (Number.isFinite(Number(recentMae))) {
    const worstThreshold = Math.max(...rules.map((r) => Number(r.holdout_mae_max) || 0));
    const driftFactor = 1 + Number(rules[0].drift_threshold ?? 0.25);
    if (Number(recentMae) > worstThreshold * driftFactor) {
      driftExceeded = true;
      driftReason = `recent MAE ${recentMae} exceeds ${(worstThreshold * driftFactor).toFixed(4)} drift threshold`;
    }
  }

  const abstain = eligible.length < minEligibleSources || driftExceeded;
  const reason = !abstain
    ? null
    : (driftExceeded
      ? driftReason
      : `eligible sources=${eligible.length} < ${minEligibleSources} (regime=${regime})`);
  return { abstain, eligible, rejected, regime, reason };
}

export async function detectRegime(client) {
  const { rows } = await client.query(
    `SELECT value
     FROM signal_history
     WHERE signal_name = 'vix' AND value_origin = 'observed'
     ORDER BY ts DESC
     LIMIT 1`,
  );
  const vix = Number(rows?.[0]?.value);
  if (Number.isFinite(vix) && vix > 30) return 'shock';
  return 'normal';
}

export async function checkEligibleSources(client, {
  targetSignal, modelVersion = 'v1', availableSources = [],
  minEligibleSources = DEFAULT_MIN_ELIGIBLE_SOURCES,
}) {
  const tableCheck = await client.query(`SELECT to_regclass('nowcast_source_eligibility') AS t`);
  if (!tableCheck.rows?.[0]?.t) {
    return { abstain: false, eligible: availableSources.map((s) => s.name), regime: 'unknown', reason: 'gate table missing, open mode' };
  }

  const { rows: rules } = await client.query(
    `SELECT source_signal, family_kind, max_lag_hours, holdout_mae_max,
            drift_threshold, regime_mask, enabled
     FROM nowcast_source_eligibility
     WHERE target_signal = $1 AND model_version = $2 AND enabled = true`,
    [targetSignal, modelVersion],
  );
  if (!rules.length) {
    return { abstain: true, eligible: [], rejected: [], regime: 'unknown', reason: `no eligibility rules for ${targetSignal}@${modelVersion}` };
  }

  const regime = await detectRegime(client);

  const { rows: driftRows } = await client.query(`
    SELECT AVG(abs_error)::float AS mae
    FROM nowcast_reconciliation
    WHERE signal_name = $1 AND reconciled_at > NOW() - INTERVAL '24 hours'
  `, [targetSignal]);
  const recentMae = driftRows?.[0]?.mae;

  return evaluateGate({ rules, regime, availableSources, recentMae, minEligibleSources });
}

File: scripts/_shared/nowcast_source_gate.py
"""
Python port of scripts/_shared/nowcast-source-gate.mjs.

Kept in lockstep with the JS implementation so that
compute-rates-nowcast.py and compute-composite-nowcasts.mjs enforce the
same abstain rules.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


DEFAULT_MIN_ELIGIBLE_SOURCES = 2


@dataclass
class GateDecision:
    abstain: bool
    eligible: list[str]
    rejected: list[dict]
    regime: str
    reason: str | None


def evaluate_gate(
    rules: list[dict],
    regime: str,
    available_sources: list[dict],
    recent_mae: float | None = None,
    min_eligible_sources: int = DEFAULT_MIN_ELIGIBLE_SOURCES,
) -> GateDecision:
    """Pure gate evaluator — no DB access. Mirror of JS evaluateGate()."""
    if not rules:
        return GateDecision(abstain=True, eligible=[], rejected=[], regime=regime, reason='no eligibility rules')

    eligible: list[str] = []
    rejected: list[dict] = []
    for src in available_sources or []:
        rule = next((r for r in rules if str(r.get('source_signal')) == str(src.get('name'))), None)
        if rule is None:
            rejected.append({'source': src.get('name'), 'reason': 'no rule'})
            continue
        regime_mask = rule.get('regime_mask') or {'normal': True, 'shock': True}
        if not regime_mask.get(regime, False):
            rejected.append({'source': src.get('name'), 'reason': f'regime={regime} disabled'})
            continue
        try:
            max_lag = float(rule.get('max_lag_hours')) if rule.get('max_lag_hours') is not None else None
        except (TypeError, ValueError):
            max_lag = None
        try:
            src_lag = float(src.get('lagHours')) if src.get('lagHours') is not None else None
        except (TypeError, ValueError):
            src_lag = None
        if src_lag is not None and max_lag is not None and src_lag > max_lag:
            rejected.append({'source': src.get('name'), 'reason': f'lag {src_lag}h > {max_lag}h'})
            continue
        eligible.append(str(src.get('name')))

    drift_exceeded = False
    drift_reason: str | None = None
    try:
        recent_mae_num = float(recent_mae) if recent_mae is not None else None
    except (TypeError, ValueError):
        recent_mae_num = None
    if recent_mae_num is not None:
        thresholds = [float(r.get('holdout_mae_max') or 0) for r in rules]
        worst = max(thresholds) if thresholds else 0.0
        drift_factor = 1 + float(rules[0].get('drift_threshold') or 0.25)
        if recent_mae_num > worst * drift_factor:
            drift_exceeded = True
            drift_reason = f'recent MAE {recent_mae_num} exceeds {worst * drift_factor:.4f} drift threshold'

    abstain = len(eligible) < min_eligible_sources or drift_exceeded
    if not abstain:
        reason = None
    elif drift_exceeded:
        reason = drift_reason
    else:
        reason = f'eligible sources={len(eligible)} < {min_eligible_sources} (regime={regime})'

    return GateDecision(abstain=abstain, eligible=eligible, rejected=rejected, regime=regime, reason=reason)


def _regclass_exists(cur, table_name: str) -> bool:
    cur.execute("SELECT to_regclass(%s) AS t", (table_name,))
    row = cur.fetchone()
    return bool(row and row[0])


def detect_regime(conn) -> str:
    with conn.cursor() as cur:
        cur.execute(
            """
            SELECT value FROM signal_history
            WHERE signal_name = 'vix' AND value_origin = 'observed'
            ORDER BY ts DESC LIMIT 1
            """
        )
        row = cur.fetchone()
    if not row:
        return 'normal'
    try:
        vix = float(row[0])
    except (TypeError, ValueError):
        return 'normal'
    return 'shock' if vix > 30 else 'normal'


def check_eligible_sources(
    conn,
    target_signal: str,
    model_version: str = 'v1',
    available_sources: list[dict] | None = None,
    min_eligible_sources: int = DEFAULT_MIN_ELIGIBLE_SOURCES,
) -> GateDecision:
    available_sources = available_sources or []
    with conn.cursor() as cur:
        if not _regclass_exists(cur, 'nowcast_source_eligibility'):
            # Gate table missing — open mode (treat sources as eligible).
            return GateDecision(
                abstain=False,
                eligible=[str(s.get('name')) for s in available_sources],
                rejected=[],
                regime='unknown',
                reason='gate table missing, open mode',
            )
        cur.execute(
            """
            SELECT source_signal, family_kind, max_lag_hours, holdout_mae_max,
                   drift_threshold, regime_mask, enabled
            FROM nowcast_source_eligibility
            WHERE target_signal = %s AND model_version = %s AND enabled = true
            """,
            (target_signal, model_version),
        )
        cols = ['source_signal', 'family_kind', 'max_lag_hours', 'holdout_mae_max',
                'drift_threshold', 'regime_mask', 'enabled']
        rules = [dict(zip(cols, row)) for row in cur.fetchall()]
        if not rules:
            return GateDecision(abstain=True, eligible=[], rejected=[], regime='unknown',
                                reason=f'no eligibility rules for {target_signal}@{model_version}')
        regime = detect_regime(conn)
        cur.execute(
            """
            SELECT AVG(abs_error)::float AS mae
            FROM nowcast_reconciliation
            WHERE signal_name = %s AND reconciled_at > NOW() - INTERVAL '24 hours'
            """,
            (target_signal,),
        )
        drift_row = cur.fetchone()
        recent_mae = drift_row[0] if drift_row else None
    return evaluate_gate(rules, regime, available_sources, recent_mae, min_eligible_sources)
